source/auth/auth.rst

```diff
@@ -1159,6 +1159,249 @@ If AWS authentication fails for any reason, the cache MUST be cleared.
+        "onOIDCRefreshToken" (defined below), if the driver supports
+        providing objects as mechanism property values.
+
+User Provided Callbacks
+```````````````````````
+
+Drivers MUST allow the user to provide callbacks for token request and
+token refresh.  The token request callback MUST accept the OIDCMechanismServerStep1 structure and return an OIDCRequestTokenResult of
+the form:
+
+.. code:: idl
+
+  OIDCRequestTokenResult:
+      description: "The result of a token request"
+      strict: false
+      fields:
+          accessToken:
+              description: "The OIDC access token"
+              type: string
+          expiresInSeconds:
+              description: "The expiration time in seconds from the current time"
+              type: int
+              optional: true
+          refreshToken:
+              description: "The OIDC refresh token"
+              type: str
+              optional: true
+
+.. code:: typescript
+
+    function onOIDCRequestToken(serverInfo: OIDCMechanismServerStep1): OIDCRequestTokenResult
+
+The driver MUST use a lock when calling a callback, and check
+for a cached result when the lock is acquired.  This is because request
+callbacks may involve human interaction, and refresh callbacks could
+use refresh tokens that can only be used once.
+
+If the callback does not return an object in the correct form of ``OIDCRequestTokenResult``, the driver MUST raise an error.   The driver will
+inspect that the correct properties are given, but MUST NOT attempt to validate
+the token(s) directly.
+
+The optional token refresh callback will accept the IDP information as
+well as the cached OIDCRequestTokenResult and return a new OIDCRequestTokenResult.
+
+
+.. code:: typescript
+
+    function onOIDCRefreshToken(serverInfo: OIDCMechanismServerStep1, tokenResult: OIDCRequestTokenResult): OIDCRequestTokenResult
+
+If the callback does not return an object in the correct form of ``OIDCRequestTokenResult``, the driver MUST raise an error.
+
+If the refresh callback is given and the request callback is not given,
+the driver MUST raise an error.  If DEVICE_NAME is given and one or more
+callbacks are given, the driver MUST raise an error.
+
+If no callbacks are given, the driver MUST enforce that a DEVICE_NAME
+mechanism_properties is set and one of ("aws", "azure", or "gcp").
+The callback mechanism can be used to support both Authentication
+Code Workflows or Device workflows that are not explicitly implemented
+by drivers.  If there is no callback and no DEVICE_NAME, or the
+DEVICE_NAME is set but credentials cannot be automatically obtained,
+the driver MUST raise an error.
+
+
+Supported Device Workflows
+``````````````````````````
+
+Drives MUST support device workflows for "aws", "azure", and "gcp", given
+by the DEVICE_NAME mechanism property.  In all cases the acquired token
+will be given as the ``jwt`` argument and the second client step of the
```

Comment:
as we discussed, the second client step in this wording means `saslStart`, not `saslContinue` (which is second step regularly). Can we describe this a bit more clear?

Reply:
Done